design/ceph/osd-migration.md

```diff
@@ -0,0 +1,106 @@
+  OSD should be migrated incase the user changes the backend store type in the spec (`spec.Storage.osdStore.type`).
+
+
+### NEW CLUSTERS
+- Currently the osd backend store is hardcoded to `bluestore` throughout the operator code.
+- Add a new environment variable called `ROOK_OSD_STORE_TYPE` in the OSD prepare and OSD pods. This will contain the OSD store type. Use this environment variable to prepare and activate the cluster.
+
+#### OSD Prepare
+- `ceph-volume prepare` command should be updated to use backend store provided in the `spec.storage.osdStore`. If no value is provided for `spec.storage.osdStore` then default to `bluestore` backend.
+
+  - **OSD on disks**:
+    - RAW MODE
+      - `ceph-volume lvm prepare --bluestore_rdr ...`
+    - LWM MODE
+      - `ceph-volume raw prepare --bluestore_rdr ...`
+  - **OSD on PVCs**:
+    -
+
+#### OSD Create
+- `OSD_STORE_FLAG` should use the value provided in the `spec.storage.osdStore`. If no values are provided by the user, then `OSD_STORE_FLAG` should default to `bluestore`
+- `ceph-volume` commands should use the `OSD_STORE_FLAG` while activating the OSDs.
+  - **OSD on disks**:
+    - RAW MODE
+    `ceph-volume raw activate --device "$DEVICE" --no-systemd --no-tmpfs`
+
+    - LVM MODE
+    `ceph-volume lvm activate --no-systemd "$OSD_STORE_FLAG" "$OSD_ID" "$OSD_UUID"`
+
+  - **OSD on PVCs**:
+
+
+
+### EXISTING CLUSTERS
+- After upgrading an existing rook ceph cluster to a version that supports `bluestore-rdr`, users can upgrade the `bluestore` based OSDs to use `bluestore-rdr`
+- Backend store of the OSDs can not be overridden, so this update will require the OSDs to be replaced one by one.
+- In order to start this osd migration, users need to patch the ceph cluster spec like below:
+
+  - Example:
+```yaml
+storage:
+  osdStore:
+    type: <desired-backend-store>
+    updateStore: yes-really-update-store
+```
+
+- Operator reconciler will perform following sequence of steps to replace an OSD:
+  1. List out all the OSDs that need to be replaced. This list will contain OSDs that have `osd-store:<osd store type>` pod label does not match the osd store type provided in `spec.storage.osdStore.type`. Lets call it `OSDDeleteList`.
+  2. If the PG status is not `active+clean`, cancel the osd replace operation and reconcile the operator.
+  3. Once the PG status is `active+clean`, delete a single OSD from the `OSDDeleteList`. Refer the deletion process mentioned below.
+  4. Continue with the existing flow of creating and upgrading the OSD, with the exception of skipping the upgrade for all the OSDs that are currently present in the `OSDDeleteList`.
+  5. Once the OSD is added back, reconcile the operator again to continue replacing the other OSDs by following the same steps mentioned above.
+
+
+
+#### DELETE OSD
+OSD can be deleted by using the following options:
+
+##### OSD PURGE JOB
```

Comment:
If we purge the OSD it will cause more data movement than if we can directly replace the OSD. Purging will cause the PGs on that OSD to be assigned to other OSDs, so the data will then be backfilled to other OSDs. When the new OSD is created, then PGs would be assigned to it and data would start backfilling to the new OSD in addition to data movement potentially to other OSDs. So if we can immediately wipe the disk and re-provision the same OSD from within the same OSD prepare job, we could avoid the potential double movement of PGs. My thoughts on this flow were in the email, but let's plan on discussing in huddle.

Reply:
marking as resolved